File: 5-Problem-A-Day/257. Binary Tree Paths.java
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode left; TreeNode
 * right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<String> binaryTreePaths(TreeNode root) {
        List<String> ans = new LinkedList<>();
        paths(root, "", ans);
        return ans;
    }

    private void paths(TreeNode root, String curr, List<String> ans) {
        if (root == null)
            return;
        if (root.left == null && root.right == null) {
            curr += Integer.toString(root.val);
            ans.add(curr);
        } else {
            paths(root.left, curr + Integer.toString(root.val) + "->", ans);
            paths(root.right, curr + Integer.toString(root.val) + "->", ans);
        }
    }
}

File: 2021/257. Binary Tree Paths.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ans;
        if (root == NULL) return ans;
        dfs(root, "", ans);
        return ans;
    }
    
    void dfs(TreeNode* root, string path, vector<string>& ans) {                
        path += to_string(root->val);
        if (root->left != NULL || root->right != NULL) {
            path += "->";
            if (root->left != NULL) {
                dfs(root->left, path, ans);                
            }
            if (root->right != NULL) {
                dfs(root->right, path, ans);
            }
        } else {
            ans.push_back(path);
        }        
    }
};